command/job_start_test.go

```diff
@@ -0,0 +1,311 @@
+	}
+
+	// stop all jobs in one command
+	args := []string{"-address", addr, "-detach"}
+	args = append(args, jobIDs...)
+	stdout, stderr, code := stop(args...)
+	must.Zero(t, code,
+		must.Sprintf("job stop stdout: %s", stdout),
+		must.Sprintf("job stop stderr: %s", stderr),
+	)
+
+	// start all jobs again in one command
+	stdout, stderr, code = start(args...)
+	must.Zero(t, code,
+		must.Sprintf("job start stdout: %s", stdout),
+		must.Sprintf("job start stderr: %s", stderr),
+	)
+
+}
+
+func TestStartCommand_MultipleCycles(t *testing.T) {
+	ci.Parallel(t)
+
+	srv, client, addr := testServer(t, true, func(c *agent.Config) {
+		c.DevMode = true
+	})
+
+	defer srv.Shutdown()
+	testutil.WaitForResult(func() (bool, error) {
+		nodes, _, err := client.Nodes().List(nil)
+		if err != nil {
+			return false, err
+		}
+		if len(nodes) == 0 {
+			return false, fmt.Errorf("missing node")
+		}
+		if _, ok := nodes[0].Drivers["mock_driver"]; !ok {
+			return false, fmt.Errorf("mock_driver not ready")
+		}
+		return true, nil
+	}, func(err error) {
+		t.Fatalf("err: %s", err)
+	})
+
+	ui := cli.NewMockUi()
+
+	job1 := testServiceJob("job-start-test")
+	job1.TaskGroups[0].Tasks[0].Resources.MemoryMB = pointer.Of(16)
+	job1.TaskGroups[0].Tasks[0].Resources.DiskMB = pointer.Of(32)
+	job1.TaskGroups[0].Tasks[0].Resources.CPU = pointer.Of(10)
+	job1.TaskGroups[0].Tasks[0].Config = map[string]interface{}{
+		"run_for": "500s",
+	}
+
+	resp, _, err := client.Jobs().Register(job1, nil)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	if code := waitForSuccess(ui, client, fullId, t, resp.EvalID); code != 0 {
+		t.Fatalf("status code non zero saw %d", code)
+	}
+
+	// wait for allocation to be running
+	allocs, _, err := client.Jobs().Allocations("job-start-test", true, nil)
+	must.NoError(t, err)
+	for _, alloc := range allocs {
+		waitForAllocRunning(t, client, alloc.ID)
+	}
+	args := []string{"-address", addr}
+	args = append(args, "job-start-test")
+	stopCmd := &JobStopCommand{Meta: Meta{Ui: ui}}
+	startCmd := &JobStartCommand{Meta: Meta{Ui: ui}}
+
+	// check multiple cycles of starting/stopping a job result in the correct version selected
+	for _ = range 3 {
+
+		code := stopCmd.Run(args)
+		must.Zero(t, code,
+			must.Sprintf("job stop stdout: %s", ui.OutputWriter.String()),
+			must.Sprintf("job stop stderr: %s", ui.ErrorWriter.String()),
+		)
+
+		code = startCmd.Run(args)
+		must.Zero(t, code,
+			must.Sprintf("job start stdout: %s", ui.OutputWriter.String()),
+			must.Sprintf("job start stderr: %s", ui.ErrorWriter.String()),
+		)
+		if newCode := waitForSuccess(ui, client, fullId, t, resp.EvalID); newCode != 0 {
+			t.Fatalf("status code non zero saw %d", newCode)
+		}
+
+		allocs, _, err = client.Jobs().Allocations("job-start-test", true, nil)
+		must.NoError(t, err)
+		waitForAllocRunning(t, client, allocs[0].ID)
+
+		// check the version selected
+		versionSelected := <-startCmd.versionCh
+		must.Eq(t, 0, versionSelected)
```

Comment:
After more investigation, we found that batch job versions do not get set as running after a stop, rather they are set to pending, and because of this, it is hard to nail down another marker to use in order for us to establish if the version we’d like to revert to is stable , so for now, this feature has been implemented and tested for service jobs only.

A key issue I noticed when testing this feature is that when stopping a job for the first time and starting it again, version 0 is set to be “running” and therefore it is selected as the version to revert to (since version 1 is the version that has stop set to true), however, when a job is stopped, version 2 (the most recent version before calling stop) is marked as “dead” so when that job is started again we will be reverting to version 0 again because it is the only running and non-stopped version available, and this will happen for all the next iterations of stop/start commands. One issue that arises is that when version zero is left behind (as after the size of the versions list reaches a limit we take the oldest version out of the versions list), we have no more running and non-stopped versions to revert to. 

For example, if I run a service job, stop it, and start it again, the job will be reverted to version 0, since that is the most recent running and non-stopped version :
<img width="760" alt="Screenshot 2024-12-05 at 12 41 49 PM 1" src="https://github.com/user-attachments/assets/1800b155-9195-4222-941c-29066762819f">
When I stop this job and start it again, version 0 will once again be selected to revert to, because version 2 has been marked as dead after the second stop.
<img width="951" alt="Screenshot 2024-12-05 at 12 42 49 PM" src="https://github.com/user-attachments/assets/6ca987c9-ffcd-470c-a883-d3e29d7a2de2">